File: NWorldPermissions/src/top/nololiyt/worldpermissions/UpdateChecker.java
package top.nololiyt.worldpermissions;

import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;

import java.io.InputStream;
import java.math.BigDecimal;
import java.net.URL;
import java.util.Scanner;

public class UpdateChecker
{
    private Plugin plugin;
    
    public UpdateChecker(Plugin plugin)
    {
        this.plugin = plugin;
    }
    
    public void checkAndLog()
    {
        Bukkit.getScheduler().runTaskAsynchronously(this.plugin, () -> {
            try (InputStream inputStream = new URL(
                    "https://yueyinqiu.github.io/NWorldPermissions/version.html")
                    .openStream(); Scanner scanner = new Scanner(inputStream))
            {
                if (scanner.hasNextBigDecimal())
                {
                    BigDecimal latest = scanner.nextBigDecimal();
                    BigDecimal current = new BigDecimal(
                            plugin.getDescription().getVersion()
                    );
                    
                    if (latest.compareTo(current) > 0)
                    {
                        plugin.getLogger().warning("Version: '" + latest.toString() + "' is available. " +
                                "And you are now using '" + current.toString() + "'. " +
                                "Visit 'https://yueyinqiu.github.io/NWorldPermissions/download' to download the newer version.");
                    }
                }
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
        });
    }
    /*
    private String readToEnd(InputStream inputStream)
    {
        Scanner scanner = new Scanner(inputStream);
        StringBuilder stringBuilder = new StringBuilder();
        while(scanner.hasNextLine())
        {
            stringBuilder.append(scanner.nextLine());
        }
        return stringBuilder.toString();
    }
    */
}

File: NWorldPermissions/src/top/nololiyt/worldpermissions/UpdateChecker.kt
package top.nololiyt.worldpermissions

import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import org.json.JSONObject

import java.net.URL

class UpdateChecker(private val plugin: Plugin)
{
    fun checkAndLog()
    {
        Bukkit.getScheduler().runTaskAsynchronously(this.plugin, Runnable {
            try
            {
                val jsonObject = JSONObject(
                    URL("https://api.github.com/repos/yueyinqiu/NWorldPermissions/releases/latest").readText());
                val tagName = jsonObject.getString("tag_name");
                
                if (!plugin.description.version.equals(tagName, true))
                {
                    val browserDownloadUrl =
                        jsonObject.getJSONArray("assets").getJSONObject(0).getString("browser_download_url");
                    
                    plugin.logger.warning(
                        "A new version: '$tagName' is available. It can be downloaded at '$browserDownloadUrl'.");
                }
                
            }
            catch (e: Exception)
            {
                e.printStackTrace();
            }
        });
    }
}